### DISCUSSION TITLE
How can I prevent whitespaces in one parser rule when I already have hidden terminal WS?

### DISCUSSION THREAD

**Discussion by jindong-zhannng** [eclipse-langium/langium#1126] [Q&A] [answered]
Hi team, I'd like to define a rule `Annotation` to express a java-like annotation syntax. Now it's like:

```
Annotation:
  '@' name=ID ('(' args+=DeclaredParameter (',' args+=DeclaredParameter)* ')')?;
```

Now since I have hidden terminal `hidden terminal WS: /\s+/;`, whitespaces between '@' and annotation name are allowed. 

My expectation is the annotation body must be followed '@' without whitespaces, and semantic model should only contains annotation body without '@' sign.

How could I achieve it?

**Comment by msujew** [comment, marked as answer]
Hey @jindong-zhannng, basically you have two options:

1. Keep it as it is now and add a validation that shows an error in case there is space between the `@` keyword and the `name` property.
2. Define a new terminal like this `terminal AnnotationName: '@' ID;` and use it as the `name` property. Make sure to override the `DefaultValueConverter` to remove the `@` character from the parsed value.

Note that I would recommend the first approach, as it makes parsing more error tolerant, and custom validations usually lead to better error messages for end users.
